pkg/loop/datasource.go

```diff
@@ -39,6 +41,13 @@ type dataSourceServer struct {
 }
 
 func (d *dataSourceServer) Observe(ctx context.Context, _ *emptypb.Empty) (*pb.ObserveReply, error) {
+	// Pipeline observations may return results after the context is cancelled, so we modify the
+	// deadline to give them time to return before the parent context deadline.
+	var cancel func()
+	ctx, cancel = utils.ContextWithDeadlineFn(ctx, func(old time.Time) time.Time {
+		return old.Add(-time.Until(old) / 10)
```

Comment:
does dividing a duration by an integer always work as expected? i have a vague memory of being tripped up by this in the past.

Reply:
`time.Duration` is nanosecond precision, so I think the point that rounding errors become problematic is already too short of a duration to do actual work anyways.